### PR TITLE
Adding attention mechanism with sampled softmax

### PR DESCRIPTION
**Pull request by segurac**
The attention mechanism needs a lot of memory when the vocabulary size is very large. Based on the translation model in Tensorflow repository https://github.com/tensorflow/tensorflow/blob/r0.11/tensorflow/models/rnn/translate/seq2seq_model.py I added the option to use "sampled softmax" as the loss function, that trains only with softmaxSamples random words at each step instead of using all possible words.

Setting softmaxSamples to zero disables sampled softmax.

### CODE REVIEW

**Comment by Conchylicultor** [comment]
Thanks, I'll try to review that this week end.
How was the results compare to the original version ?

**Comment by segurac** [comment]
It trained much faster and with lower memory requirements. I am using a vocabulary size of 40k words, which is impossible to train with the attention mechanism without using sampled softmax o a similar technique.

Regarding the results, the answers seem to be more grammatically and syntactically correct, and it seems to focus on particular words of the sentence. 

I also have implemented other features, like sampled decoding, instead of the greedy max decoding, and creating a seq2seq model that shares the parameters of the embedding at the encoder and the decoder. I'll let you know when they are ready, I don want to diverge too much from your code!

**Comment by dimeldo** [comment]
@segurac Can you share the results you got (conversations) and the parameters you've used?

**Comment by segurac** [comment]
The parameters:
--hiddenSize 256 --numLayers 2 --embeddingSize 128 --batchSize 128 --softmaxSamples 512

Regarding the results, I am using the spanish OpenSubtitle database from http://opus.lingfil.uu.se/OpenSubtitles2016alt.php (es.raw.tar.gz), assuming every consecutive 2 lines as a QA. I have some quick and dirty code in my repository to prepare the db, but I would like to clean it up and prepare a proper pull request eventually, so that you can try it yourself.

Btw, I see that there are starting to appear many forks of the DeepQA, with very interesting features like using twitter data (https://github.com/kootenpv/DeepQA) or a dockerized chatbot (https://github.com/kootenpv/DeepQA).

**Comment by Conchylicultor** [comment]
Thank you for your contribution.

On my computer, the sampled softmax was about 2/3 times slower so I deactivated it by default. I also did some refactoring to be more aligned with the rest of the program.
